Feature: add application filter to `azuread_conditional_access_policy`

BrendanThompson opened the pull request:
## Description

This PR adds the ability to have an application filter on `azuread_conditional_access_policy`.

```
resource "azuread_conditional_access_policy" "this" {
  display_name = "access-policy"
  state        = "disabled"

  conditions {
    client_app_types    = ["other"]

    applications {
      filter {
        mode = "include"
        rule = "CustomSecurityAttribute.AttributeName -eq \"AttributeValue\""
      }
    }

    users {
      included_roles = ["Contributor"]
    }
  }
}
```

## Change Log

- `azuread_conditional_access_policy` – add the `filter` attribute to the `applications` block allowing for applications to be filtered [GH-1357]

## Related Issue(s)

Fixes #1318

iwarapter committed:
fixed #1031 adds provision on demand

KenSpur committed:
azuread_application_pre_authorized: fix for concurrency bug during destroy

Co-authored-by: KenSpur <spur.ken@gmail.com>

JonasBak committed:
feat: allow data.azuread_application lookup using identifier_uris

iwarapter committed:
feat: add runtime provider debugging capability

BrendanThompson committed:
feat: add filter to applications in schema

Signed-off-by: Brendan Thompson <github@blt.is>

BrendanThompson committed:
feat: add expansion for applications filter

Signed-off-by: Brendan Thompson <github@blt.is>

BrendanThompson committed:
feat: make flatten filter generic and use for device and app

Signed-off-by: Brendan Thompson <github@blt.is>

BrendanThompson committed:
chore: ensure only one inclusion option is used

Signed-off-by: Brendan Thompson <github@blt.is>

BrendanThompson commented:
@tombuildsstuff — don't suppose you've time to give some feedback on this PR as well?

manicminer reviewed (changes requested):
Thanks @TomasKunka for this addition.

This is looking great - if you could add an acceptance test case that uses a filter for an application condition then this should be good to merge. Conditional Access Policies have shown to yield odd payload parsing bugs further down the line so having coverage is necessary to give us a good shot at spotting regressions.

Thanks!

hashicorp-copywrite[bot] committed:
[COMPLIANCE] Add Copyright and License Headers

manicminer committed:
Changelog for #1299

manicminer committed:
azuread_group: support `SkipExchangeInstantOn` for `behaviors`. Fixes #1315

iwarapter committed:
add triggers arguement to allow re-triggering use cases

manicminer committed:
Changelog for #1370

manicminer committed:
linting, tidying

manicminer committed:
bugfix: synchronization package registration oopsie

manicminer committed:
make generate

manicminer committed:
Changlog for #1032

manicminer committed:
v2.49.0

BrendanThompson commented:
@manicminer thanks for reviewing this, I will try and get an acceptance test in today 😃

manicminer commented:
Great thanks, and sorry for the mis-tag! 😊

audunsolemdal committed:
Fix example code for azuread_group_role_management_policy

audunsolemdal committed:
Fix typo

audunsolemdal committed:
Add example for required_conditional_access_authentication_context

manicminer committed:
bugfix: resolve several potential crashes in new PIM resources

manicminer committed:
another crash fix

hashicorp-tsccr[bot] committed:
Result of tsccr-helper -log-level=info gha update .

manicminer committed:
Changelog for #1375

manicminer committed:
v2.49.1

JonasBak committed:
fix: use identifier_uri instead of identifier_uris for lookup

hashicorp-copywrite[bot] committed:
[COMPLIANCE] Add Copyright and License Headers

BrendanThompson commented:
@manicminer — sorry it's taken me a few days to get back to you. I started writing the tests then I realised that we will actually require some attributes to be set on wherever the Acceptance Tests are run, otherwise the tests will fail. There is currently also no mechanism to deploy said attributes with Terraform ☹️.

https://learn.microsoft.com/en-us/entra/fundamentals/custom-security-attributes-overview

The above link are the attributes we need in-place. How would you like to proceed on this, if possible were you able to add an attribute and I can test with that. Then perhaps the next thing is to work on integrating the attribute creation into the provider? (Happy to work in this)

BrendanThompson commented:
I cannot think of any way to test it properly without the custom attributes being there to be honest. And testing it without an attribute would yield odd results due to the API. We could add a notice on the resource itself, like a warning around it. _Buyer beware_ until the security attributes are available. What are your thoughts?

manicminer commented:
I would also love to ship this but unless we are able to demonstrate some level of testing, we unfortunately won't be able to add a property or block to a resource without it, which I guess means this will have to wait until we have a resource to support custom security attributes.

manicminer committed:
bugfix: run d.Partial() to avoid setting invalid password to state when a user update fails

manicminer committed:
Update docs/data-sources/application.md

manicminer committed:
azuread_user: acceptance test for setting an invalid password

manicminer committed:
CHangelog for #1303

dlaguerta committed:
use single runner label for custom-linux-large

BrendanThompson commented:
Hey @manicminer,

I have opened a PR in `Hamilton` (https://github.com/manicminer/hamilton/pull/281) to add support for both `attribute sets` and `custom security attribute definitions` once that has gone through I will add support for them in the provider and update the tests here 😃

manicminer committed:
data.azuread_service_principal: display name comparison should be case-insensitive

manicminer committed:
Changelog for #1381

manicminer committed:
Changelog for #1308

manicminer committed:
azuread_conditional_access_policy: improve handling of the `session_controls` block

- Make `sign_in_frequency_authentication_type` and
  `sign_in_frequency_internal` both Optional + Computed and remove their
  default values.
- Handle the setting of these default values in the
  `expandConditionalAccessSessionControls()` function.
- Expand test coverage to all reasonable permutatons of the properties
  in this block to ensure no trailing diff or incorrect setting of
  values in the request.

manicminer committed:
dependencies: updating to `v0.68.0` of `github.com/manicminer/hamilton`

manicminer committed:
Changelog for #1382

manicminer committed:
typo fix

manicminer committed:
v2.50.0

manicminer committed:
tooling: updating version of hashicorp/ghaction-terraform-provider-release to v4.0.1

manicminer committed:
tooling: enable running a release from a branch instead of `main`

hashicorp-tsccr[bot] committed:
Result of tsccr-helper -log-level=info gha update .

manicminer committed:
dependencies: updating to `v0.70.0` of `github.com/manicminer/hamilton`

manicminer committed:
bugfix: allow disabling review settings for `azuread_access_package_assignment_policy`

manicminer committed:
docs: add note on using `ignore_changes` on `administrative_unit_ids` for `azuread_group`

manicminer committed:
test fixes for administrative units

manicminer committed:
Changelog for #1394

manicminer commented:
@BrendanThompson The hamilton version has been updated in the provider, hopefully this unblocks your work on this 🙂

manicminer committed:
data.azuread_users: support for the `mails` property

manicminer committed:
Changelog for #1400

manicminer committed:
v2.51.0

manicminer committed:
azuread_application: support `ignore_changes` lifecycle argument for `app_role`, `oauth2_permission_scope`, `identifier_uris`, `optional_claims`, and `required_resource_access`

manicminer committed:
azuread_application: basicFromTemplate test fix

manicminer committed:
azuread_application: basicFromTemplate more test fix

manicminer committed:
internal concurrency lock when updating azuread_application / azuread_application_registration

manicminer committed:
azuread_application: work around very buggy API when instantiating from template

The `/instantiate` operation can return a 404 whlst also processing the
request completely normally. This leads to orphaned objects in the
directory, and a resource that cannot successfully Create.

Work around this by polling for the application and service principal
that you'd expect to see created out-of-band, whenever a 404 is
received. Also set a temporary `displayName` for the application, as
this is the only means we have to identify the resulting object is the
one we are looking for.

Unfortunately this means that this workaround cannot be implemented for
the `azuread_application_from_template` resource, since that resource
intentionally avoids changing the implicitly created
`user_impersonation` scope - this will get created with nonsensical
display names / descriptions in the consent flow. Since the whole point
of the standalone resource is to inherit scopes from the template, this
makes it infeasible to add this workaround there.

manicminer committed:
REBASE! dependencies: updating to `REBASE ME!` of `github.com/manicminer/hamilton` - depends on https://github.com/manicminer/hamilton/pull/285

manicminer committed:
Changelog for #1403

manicminer committed:
Changelog for #1406

manicminer committed:
v2.52.0

jamie-oconnell commented:
Are there any current blockers to merge this? Happy holidays!

bitcarousel commented:
It seems to be almost a year ago since Feature created. Will this make it into a release soon would be great?

jackofallops reviewed (changes requested):
Hi @BrendanThompson - Not sure exactly what's happened with this PR, but can you rebase on latest main to see if we can get rid of all the unrelated "touched" files - we can't review/accept this PR as it stands with ~90 unrelated files modified in it. Sorry.

Thanks

bitcarousel commented:
Will this ever be merged into main?
I have honestly been waiting almost two years to finally be able to use application filters on Conditional Access policies with the azuread provider. I don’t think this is an edge case anymore.

bubbletroubles commented:
I think #1780 might be more appropriate to merge, its cleaner from a git history perspective and does the same thing. @jackofallops  can someone take a look?

sreallymatt commented:
superseded by #1780
